# Spotify Top 50 Tracks of 2020 EDA

In this notebook, we will explore the Spotify Top 50 Tracks of 2020 dataset.

EDA (Exploratory Data Analysis) means looking through the data before making conclusions.

## Project Goal

The goal of this project is to analyze the Spotify Top 50 Tracks of 2020 dataset and understand what artists, genres, and audio features appear most often among popular songs.

## Dataset Description

This dataset contains 50 popular Spotify tracks from 2020. Each row represents one song. The columns include song information such as artist, album, track name, genre, and audio features like danceability, energy, loudness, tempo, and duration.

## Phase 2: Load The Dataset

First, we import Pandas and load the CSV file.

- `import pandas as pd` loads the Pandas library.
- `pd.read_csv(...)` reads a CSV file into a table-like object called a DataFrame.
- A DataFrame is like a spreadsheet inside Python: it has rows and columns.

In [1]:
import pandas as pd

csv_path = "../data/spotifytoptracks.csv"
df = pd.read_csv(csv_path)

df.head()

,Unnamed: 0,artist,album,track_name,track_id,energy,danceability,key,loudness,acousticness,speechiness,instrumentalness,liveness,valence,tempo,duration_ms,genre
0,0,The Weeknd,After Hours,Blinding Lights,0VjIjW4GlUZAMYd2vXMi3b,0.730,0.514,1,-5.934,0.00146,0.0598,0.000095,0.0897,0.334,171.005,200040,R&B/Soul
1,1,Tones And I,Dance Monkey,Dance Monkey,1rgnBhdG2JDFTbYkYRZAku,0.593,0.825,6,-6.401,0.68800,0.0988,0.000161,0.1700,0.540,98.078,209755,Alternative/Indie
2,2,Roddy Ricch,Please Excuse Me For Being Antisocial,The Box,0nbXyq5TXYPCO7pr3N8S4I,0.586,0.896,10,-6.687,0.10400,0.0559,0.000000,0.7900,0.642,116.971,196653,Hip-Hop/Rap
3,3,SAINt JHN,Roses (Imanbek Remix),Roses - Imanbek Remix,2Wo6QQD1KMDWeFkkjLqwx5,0.721,0.785,8,-5.457,0.01490,0.0506,0.004320,0.2850,0.894,121.962,176219,Dance/Electronic
4,4,Dua Lipa,Future Nostalgia,Don't Start Now,3PfIrDoz19wz7qK7tYeu62,0.793,0.793,11,-4.521,0.01230,0.0830,0.000000,0.0951,0.679,123.950,183290,Nu-disco


## Basic Size Check

Before cleaning or analyzing, we check the size of the dataset.

- Rows usually mean records or observations. Here, each row is one track.
- Columns usually mean features or variables. Here, columns describe track details such as artist, album, tempo, and genre.

In [2]:
df.shape

(50, 17)

## Column Names

Column names tell us what information the dataset contains.

In [3]:
df.columns

Index(['Unnamed: 0', 'artist', 'album', 'track_name', 'track_id', 'energy',
       'danceability', 'key', 'loudness', 'acousticness', 'speechiness',
       'instrumentalness', 'liveness', 'valence', 'tempo', 'duration_ms',
       'genre'],
      dtype='object')

## Data Types

A data type tells Python what kind of value is stored in each column.

- `object` usually means text.
- `int64` means whole numbers.
- `float64` means decimal numbers.

In [4]:
df.dtypes

Unnamed: 0            int64
artist               object
album                object
track_name           object
track_id             object
energy              float64
danceability        float64
key                   int64
loudness            float64
acousticness        float64
speechiness         float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
duration_ms           int64
genre                object
dtype: object

## Phase 3: Observations And Features

Now we count observations and features.

- An **observation** is one row in the dataset. Here, one observation means one song.
- A **feature** is one column in the dataset. Here, one feature means one detail about a song, such as artist, tempo, or genre.

In [5]:
number_of_observations = df.shape[0]
number_of_features = df.shape[1]

print("Number of observations:", number_of_observations)
print("Number of features:", number_of_features)

Number of observations: 50
Number of features: 17


## Categorical And Numeric Features

A **categorical feature** stores labels or groups, like artist names or genres.

A **numeric feature** stores numbers that we can calculate with, like tempo, loudness, or duration.

In [6]:
categorical_features = df.select_dtypes(include="object").columns.tolist()
numeric_features = df.select_dtypes(include="number").columns.tolist()

print("Categorical features:")
print(categorical_features)

print("\nNumeric features:")
print(numeric_features)

Categorical features:
['artist', 'album', 'track_name', 'track_id', 'genre']

Numeric features:
['Unnamed: 0', 'energy', 'danceability', 'key', 'loudness', 'acousticness', 'speechiness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'duration_ms']


### Quick Note About `Unnamed: 0`

`Unnamed: 0` looks like an old row number that was saved into the CSV file. It probably does not describe the songs themselves, so we will treat it as a cleanup item in the next phase.

In [7]:
df.isnull().sum()

Unnamed: 0          0
artist              0
album               0
track_name          0
track_id            0
energy              0
danceability        0
key                 0
loudness            0
acousticness        0
speechiness         0
instrumentalness    0
liveness            0
valence             0
tempo               0
duration_ms         0
genre               0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df["Unnamed: 0"].head()

0    0
1    1
2    2
3    3
4    4
Name: Unnamed: 0, dtype: int64

In [10]:
df = df.drop(columns=["Unnamed: 0"])
df.head()

,artist,album,track_name,track_id,energy,danceability,key,loudness,acousticness,speechiness,instrumentalness,liveness,valence,tempo,duration_ms,genre
0,The Weeknd,After Hours,Blinding Lights,0VjIjW4GlUZAMYd2vXMi3b,0.730,0.514,1,-5.934,0.00146,0.0598,0.000095,0.0897,0.334,171.005,200040,R&B/Soul
1,Tones And I,Dance Monkey,Dance Monkey,1rgnBhdG2JDFTbYkYRZAku,0.593,0.825,6,-6.401,0.68800,0.0988,0.000161,0.1700,0.540,98.078,209755,Alternative/Indie
2,Roddy Ricch,Please Excuse Me For Being Antisocial,The Box,0nbXyq5TXYPCO7pr3N8S4I,0.586,0.896,10,-6.687,0.10400,0.0559,0.000000,0.7900,0.642,116.971,196653,Hip-Hop/Rap
3,SAINt JHN,Roses (Imanbek Remix),Roses - Imanbek Remix,2Wo6QQD1KMDWeFkkjLqwx5,0.721,0.785,8,-5.457,0.01490,0.0506,0.004320,0.2850,0.894,121.962,176219,Dance/Electronic
4,Dua Lipa,Future Nostalgia,Don't Start Now,3PfIrDoz19wz7qK7tYeu62,0.793,0.793,11,-4.521,0.01230,0.0830,0.000000,0.0951,0.679,123.950,183290,Nu-disco


The column `Unnamed: 0` only repeated the row numbers from the CSV file. It did not describe the songs, so I removed it from the dataset.

In [11]:
df.describe()

,energy,danceability,key,loudness,acousticness,speechiness,instrumentalness,liveness,valence,tempo,duration_ms
count,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000,50.000000
mean,0.609300,0.716720,5.720000,-6.225900,0.256206,0.124158,0.015962,0.196552,0.555710,119.690460,199955.360000
std,0.154348,0.124975,3.709007,2.349744,0.265250,0.116836,0.094312,0.176610,0.216386,25.414778,33996.122488
min,0.225000,0.351000,0.000000,-14.454000,0.001460,0.029000,0.000000,0.057400,0.060500,75.801000,140526.000000
25%,0.494000,0.672500,2.000000,-7.552500,0.052800,0.048325,0.000000,0.093950,0.434000,99.557250,175845.500000
50%,0.597000,0.746000,6.500000,-5.991500,0.188500,0.070050,0.000000,0.111000,0.560000,116.969000,197853.500000
75%,0.729750,0.794500,8.750000,-4.285500,0.298750,0.155500,0.000020,0.271250,0.726250,132.317000,215064.000000
max,0.855000,0.935000,11.000000,-3.280000,0.934000,0.487000,0.657000,0.792000,0.925000,180.067000,312820.000000


In [12]:
df["duration_minutes"] = df["duration_ms"] / 1000 / 60
df[["track_name", "artist", "duration_ms", "duration_minutes"]].head()

,track_name,artist,duration_ms,duration_minutes
0,Blinding Lights,The Weeknd,200040,3.334000
1,Dance Monkey,Tones And I,209755,3.495917
2,The Box,Roddy Ricch,196653,3.277550
3,Roses - Imanbek Remix,SAINt JHN,176219,2.936983
4,Don't Start Now,Dua Lipa,183290,3.054833


The numeric values look reasonable. I did not remove any outliers because the minimum and maximum values still make sense for songs in this dataset. I added `duration_minutes` to make song length easier to understand.

### Artist Counts

In [13]:
df["artist"].value_counts()

artist
Billie Eilish      3
Dua Lipa           3
Travis Scott       3
Justin Bieber      2
Harry Styles       2
Lewis Capaldi      2
Post Malone        2
Cardi B            1
Surfaces           1
Eminem             1
BTS                1
BENEE              1
Surf Mesa          1
Lady Gaga          1
The Weeknd         1
Shawn Mendes       1
Regard             1
Black Eyed Peas    1
THE SCOTTS         1
Bad Bunny          1
Juice WRLD         1
Ariana Grande      1
Maluma             1
Topic              1
24kGoldn           1
Tones And I        1
Jawsh 685          1
Future             1
Maroon 5           1
Doja Cat           1
Drake              1
Lil Mosey          1
Arizona Zervas     1
KAROL G            1
Trevor Daniel      1
Powfu              1
DaBaby             1
SAINt JHN          1
Roddy Ricch        1
JP Saxe            1
Name: count, dtype: int64

Billie Eilish, Dua Lipa, and Travis Scott appear most often in this Top 50 dataset, with 3 songs each. This suggests that these artists had especially strong chart presence in 2020.

In [14]:
df["artist"].nunique()

40

There are 40 unique artists in the Top 50 dataset. Since there are 50 songs total, most artists appear once, while a few artists appear multiple times.

### Album Counts

In [15]:
df["album"].nunique()

45

In [16]:
df["album"].value_counts()

album
Future Nostalgia                                              3
Hollywood's Bleeding                                          2
Fine Line                                                     2
Changes                                                       2
Ride It                                                       1
Where the Light Is                                            1
Music To Be Murdered By                                       1
Dynamite (DayTime Version)                                    1
Hey u x                                                       1
ily (i love you baby) (feat. Emilee)                          1
Chromatica                                                    1
HIGHEST IN THE ROOM                                           1
PAPI JUANCHO                                                  1
After Hours                                                   1
WHEN WE ALL FALL ASLEEP, WHERE DO WE GO?                      1
Birds In The Trap Sing McKnight   

There are 45 unique albums in the Top 50 dataset. Most albums appear once, but `Future Nostalgia` appears 3 times. `Hollywood's Bleeding`, `Fine Line`, and `Changes` each appear 2 times. This suggests that a few albums produced multiple very popular tracks.

### Genre Counts

In [17]:
df["genre"].value_counts()

genre
Pop                                   14
Hip-Hop/Rap                           13
Dance/Electronic                       5
Alternative/Indie                      4
R&B/Soul                               2
 Electro-pop                           2
Nu-disco                               1
R&B/Hip-Hop alternative                1
Pop/Soft Rock                          1
Pop rap                                1
Hip-Hop/Trap                           1
Dance-pop/Disco                        1
Disco-pop                              1
Dreampop/Hip-Hop/R&B                   1
Alternative/reggaeton/experimental     1
Chamber pop                            1
Name: count, dtype: int64

In [18]:
df["genre"] = df["genre"].str.strip()
df["genre"].value_counts()

genre
Pop                                   14
Hip-Hop/Rap                           13
Dance/Electronic                       5
Alternative/Indie                      4
R&B/Soul                               2
Electro-pop                            2
Nu-disco                               1
R&B/Hip-Hop alternative                1
Pop/Soft Rock                          1
Pop rap                                1
Hip-Hop/Trap                           1
Dance-pop/Disco                        1
Disco-pop                              1
Dreampop/Hip-Hop/R&B                   1
Alternative/reggaeton/experimental     1
Chamber pop                            1
Name: count, dtype: int64

Pop is the most common genre in the Top 50 dataset, with 14 songs. Hip-Hop/Rap is close behind with 13 songs. This suggests that mainstream pop and rap were especially strong in the 2020 Top 50.
I noticed that `Electro-pop` had an extra space at the beginning in the original data. I used `.str.strip()` to remove extra spaces from the genre names before making final genre counts.

### Most and Least Danceable Songs

In [19]:
df[["track_name", "artist", "danceability"]].sort_values(by="danceability", ascending=False).head(10)

,track_name,artist,danceability
27,WAP (feat. Megan Thee Stallion),Cardi B,0.935
2,The Box,Roddy Ricch,0.896
39,Ride It,Regard,0.880
28,Sunday Best,Surfaces,0.878
33,Supalonely (feat. Gus Dapperton),BENEE,0.862
40,goosebumps,Travis Scott,0.841
49,SICKO MODE,Travis Scott,0.834
15,Toosie Slide,Drake,0.830
1,Dance Monkey,Tones And I,0.825
29,Godzilla (feat. Juice WRLD),Eminem,0.808


`WAP (feat. Megan Thee Stallion)` by Cardi B has the highest danceability score at 0.935. Several Hip-Hop/Rap tracks appear in the top danceability results, which suggests that rhythm-focused tracks were among the most danceable songs in this dataset.

In [20]:
df[["track_name", "artist", "danceability"]].sort_values(by="danceability", ascending=True).head(10)

,track_name,artist,danceability
44,lovely (with Khalid),Billie Eilish,0.351
16,Before You Go,Lewis Capaldi,0.459
47,If the World Was Ending - feat. Julia Michaels,JP Saxe,0.464
9,Someone You Loved,Lewis Capaldi,0.501
45,Lucid Dreams,Juice WRLD,0.511
0,Blinding Lights,The Weeknd,0.514
6,Watermelon Sugar,Harry Styles,0.548
46,Stuck with U (with Justin Bieber),Ariana Grande,0.597
36,HIGHEST IN THE ROOM,Travis Scott,0.598
43,Safaera,Bad Bunny,0.607


`lovely (with Khalid)` by Billie Eilish has the lowest danceability score at 0.351. The least danceable songs include several slower or more emotional tracks, which shows that a song does not need to be highly danceable to become popular.

### Loudest and Quietest Songs

In [21]:
df[["track_name", "artist", "loudness"]].sort_values(by="loudness", ascending=False).head(10)

,track_name,artist,loudness
10,Tusa,KAROL G,-3.280
40,goosebumps,Travis Scott,-3.370
31,Break My Heart,Dua Lipa,-3.434
38,Hawái,Maluma,-3.454
12,Circles,Post Malone,-3.497
23,Mood (feat. iann dior),24kGoldn,-3.558
21,Adore You,Harry Styles,-3.675
49,SICKO MODE,Travis Scott,-3.714
48,Physical,Dua Lipa,-3.756
35,Rain On Me (with Ariana Grande),Lady Gaga,-3.764


`Tusa` by KAROL G is the loudest song in the dataset, with a loudness value of -3.280. Loudness values are negative, so values closer to zero are louder.

In [22]:
df[["track_name", "artist", "loudness"]].sort_values(by="loudness", ascending=True).head(10)

,track_name,artist,loudness
24,everything i wanted,Billie Eilish,-14.454
26,bad guy,Billie Eilish,-10.965
44,lovely (with Khalid),Billie Eilish,-10.109
47,If the World Was Ending - feat. Julia Michaels,JP Saxe,-10.086
15,Toosie Slide,Drake,-8.820
7,death bed (coffee for your head),Powfu,-8.765
36,HIGHEST IN THE ROOM,Travis Scott,-8.764
8,Falling,Trevor Daniel,-8.756
20,Savage Love (Laxed - Siren Beat),Jawsh 685,-8.520
5,ROCKSTAR (feat. Roddy Ricch),DaBaby,-7.956


`everything i wanted` by Billie Eilish is the quietest song in the dataset, with a loudness value of -14.454. Billie Eilish appears several times in the quietest songs list, suggesting that quieter or softer production can still perform very well in the Top 50.

### Longest and Shortest Songs

In [23]:
df[["track_name", "artist", "duration_minutes"]].sort_values(by="duration_minutes", ascending=False).head(10)

,track_name,artist,duration_minutes
49,SICKO MODE,Travis Scott,5.213667
43,Safaera,Bad Bunny,4.919617
15,Toosie Slide,Drake,4.117650
24,everything i wanted,Billie Eilish,4.090433
40,goosebumps,Travis Scott,4.063950
45,Lucid Dreams,Juice WRLD,3.997267
19,Life Is Good (feat. Drake),Future,3.965300
17,Say So,Doja Cat,3.964883
46,Stuck with U (with Justin Bieber),Ariana Grande,3.808033
33,Supalonely (feat. Gus Dapperton),BENEE,3.724800


`SICKO MODE` by Travis Scott is the longest song in the dataset at about 5.21 minutes. Most of the longest songs are between about 3.7 and 5.2 minutes.

In [24]:
df[["track_name", "artist", "duration_minutes"]].sort_values(by="duration_minutes", ascending=True).head(10)

,track_name,artist,duration_minutes
23,Mood (feat. iann dior),24kGoldn,2.342100
37,Sunflower - Spider-Man: Into the Spider-Verse,Post Malone,2.626000
39,Ride It,Regard,2.626767
28,Sunday Best,Surfaces,2.642850
8,Falling,Trevor Daniel,2.656367
13,Blueberry Faygo,Lil Mosey,2.709117
11,ROXANNE,Arizona Zervas,2.727267
42,THE SCOTTS,THE SCOTTS,2.766300
22,Breaking Me,Topic,2.779900
20,Savage Love (Laxed - Siren Beat),Jawsh 685,2.856250


`Mood (feat. iann dior)` by 24kGoldn is the shortest song in the dataset at about 2.34 minutes. The shortest Top 50 songs are all under 3 minutes, which suggests that short songs can still perform very strongly.

### Genre Feature Comparison

In [25]:
df.groupby("genre")[["danceability", "energy", "loudness", "duration_minutes"]].mean().sort_values(by="danceability", ascending=False)

,danceability,energy,loudness,duration_minutes
genre,,,,
Hip-Hop/Trap,0.935000,0.454000,-7.509000,3.125683
Nu-disco,0.793000,0.793000,-4.521000,3.054833
Electro-pop,0.789500,0.475000,-8.898500,2.938825
R&B/Hip-Hop alternative,0.784000,0.430000,-8.756000,2.656367
Hip-Hop/Rap,0.765538,0.582615,-6.917846,3.388987
Dance/Electronic,0.755000,0.762000,-5.338000,2.903317
Dreampop/Hip-Hop/R&B,0.755000,0.522000,-4.368000,2.626000
Disco-pop,0.746000,0.765000,-4.410000,3.317567
Dance-pop/Disco,0.730000,0.729000,-3.434000,3.697000


This table compares average audio features by genre. However, some genres only have one song, so their averages should be interpreted carefully. A genre with one song is not enough to represent the whole genre.

In [26]:
genre_counts = df["genre"].value_counts()
genres_with_multiple_songs = genre_counts[genre_counts >= 2].index

df[df["genre"].isin(genres_with_multiple_songs)].groupby("genre")[["danceability", "energy", "loudness", "duration_minutes"]].mean().sort_values(by="danceability", ascending=False)

,danceability,energy,loudness,duration_minutes
genre,,,,
Electro-pop,0.789500,0.475000,-8.898500,2.938825
Hip-Hop/Rap,0.765538,0.582615,-6.917846,3.388987
Dance/Electronic,0.755000,0.762000,-5.338000,2.903317
Pop,0.677571,0.587286,-6.460357,3.455290
Alternative/Indie,0.661750,0.551000,-5.421000,3.460463
R&B/Soul,0.650500,0.701500,-5.255500,3.649442


After filtering to genres with at least 2 songs, Electro-pop has the highest average danceability. Dance/Electronic has the highest average energy. This comparison is more useful than the full genre table because it avoids relying too much on genres that only appear once.

### Correlation Analysis

In [27]:
numeric_columns = ["energy", "danceability", "loudness", "acousticness", "speechiness", "instrumentalness", "liveness", "valence", "tempo", "duration_minutes"]

df[numeric_columns].corr()

,energy,danceability,loudness,acousticness,speechiness,instrumentalness,liveness,valence,tempo,duration_minutes
energy,1.000000,0.152552,0.791640,-0.682479,0.074267,-0.385515,0.069487,0.393453,0.075191,0.081971
danceability,0.152552,1.000000,0.167147,-0.359135,0.226148,-0.017706,-0.006648,0.479953,0.168956,-0.033763
loudness,0.791640,0.167147,1.000000,-0.498695,-0.021693,-0.553735,-0.069939,0.406772,0.102097,0.064130
acousticness,-0.682479,-0.359135,-0.498695,1.000000,-0.135392,0.352184,-0.128384,-0.243192,-0.241119,-0.010988
speechiness,0.074267,0.226148,-0.021693,-0.135392,1.000000,0.028948,-0.142957,0.053867,0.215504,0.366976
instrumentalness,-0.385515,-0.017706,-0.553735,0.352184,0.028948,1.000000,-0.087034,-0.203283,0.018853,0.184709
liveness,0.069487,-0.006648,-0.069939,-0.128384,-0.142957,-0.087034,1.000000,-0.033366,0.025457,-0.090188
valence,0.393453,0.479953,0.406772,-0.243192,0.053867,-0.203283,-0.033366,1.000000,0.045089,-0.039794
tempo,0.075191,0.168956,0.102097,-0.241119,0.215504,0.018853,0.025457,0.045089,1.000000,0.130328
duration_minutes,0.081971,-0.033763,0.064130,-0.010988,0.366976,0.184709,-0.090188,-0.039794,0.130328,1.000000


The strongest relationship in this table is between energy and loudness, with a correlation of about 0.79. This means songs with higher energy also tend to be louder.

Energy and acousticness have a negative correlation of about -0.68. This means more acoustic songs tend to have lower energy in this dataset.

Danceability and valence have a moderate positive correlation of about 0.48. Valence describes how positive or cheerful a song sounds, so this suggests that more danceable songs also tend to sound more positive.

Correlation does not prove cause and effect. It only shows whether two numeric features tend to move together.

## Final Summary

This analysis explored the Spotify Top 50 Tracks of 2020 dataset.

Main findings:

- Billie Eilish, Dua Lipa, and Travis Scott appeared most often, with 3 songs each.
- Pop was the most common genre, followed closely by Hip-Hop/Rap.
- `WAP (feat. Megan Thee Stallion)` by Cardi B had the highest danceability score.
- `Tusa` by KAROL G was the loudest song.
- `SICKO MODE` by Travis Scott was the longest song.
- Energy and loudness had a strong positive relationship, meaning higher-energy songs tended to also be louder.

Overall, popular songs in this dataset were often pop or rap tracks, but the Top 50 included a variety of styles, song lengths, and audio features.


## Limitations

This dataset only contains 50 songs from one year, so the results should not be treated as rules for all Spotify music. The findings describe this specific Top 50 dataset only.

## Assignment Requirement Checks

This section answers the remaining assignment questions directly. Some of these questions were explored earlier, but here they are shown in a clear checklist style for review.

### Duplicate Features Check

The assignment asks us to remove duplicate samples and features. Earlier, duplicate rows were checked with `df.duplicated().sum()`. Here, I also check whether any columns are duplicated.

In [28]:
duplicate_feature_names = df.columns[df.columns.duplicated()].tolist()

columns = df.columns.tolist()
duplicate_feature_contents = []

for i in range(len(columns)):
    for j in range(i + 1, len(columns)):
        if df[columns[i]].equals(df[columns[j]]):
            duplicate_feature_contents.append((columns[i], columns[j]))

print("Duplicate feature names:", duplicate_feature_names)
print("Duplicate feature contents:", duplicate_feature_contents)

Duplicate feature names: []
Duplicate feature contents: []


There are no duplicate feature names and no columns with identical contents. This means no duplicate features needed to be removed.

### Outlier Review

Outliers are unusually high or low values. I use the IQR method to flag possible outliers, then review whether they make sense for this dataset.

In [29]:
outlier_columns = [
    "energy", "danceability", "loudness", "acousticness", "speechiness",
    "instrumentalness", "liveness", "valence", "tempo", "duration_minutes"
]

outlier_summary = {}

for column in outlier_columns:
    q1 = df[column].quantile(0.25)
    q3 = df[column].quantile(0.75)
    iqr = q3 - q1
    lower_limit = q1 - 1.5 * iqr
    upper_limit = q3 + 1.5 * iqr
    outlier_count = df[(df[column] < lower_limit) | (df[column] > upper_limit)].shape[0]
    outlier_summary[column] = outlier_count

pd.Series(outlier_summary, name="possible_outliers")

energy               0
danceability         3
loudness             1
acousticness         7
speechiness          6
instrumentalness    12
liveness             3
valence              0
tempo                0
duration_minutes     2
Name: possible_outliers, dtype: int64

Some columns may have possible outliers, but these values still make sense for music data. For example, a song can be much quieter, more acoustic, or more instrumental than other popular tracks. I did not remove these rows because they appear to be valid songs, not data errors.

### Tracks With Danceability Above 0.7

In [30]:
danceability_above_07 = df[df["danceability"] > 0.7][["track_name", "artist", "danceability"]].sort_values(by="danceability", ascending=False)
danceability_above_07

,track_name,artist,danceability
27,WAP (feat. Megan Thee Stallion),Cardi B,0.935
2,The Box,Roddy Ricch,0.896
39,Ride It,Regard,0.880
28,Sunday Best,Surfaces,0.878
33,Supalonely (feat. Gus Dapperton),BENEE,0.862
40,goosebumps,Travis Scott,0.841
49,SICKO MODE,Travis Scott,0.834
15,Toosie Slide,Drake,0.830
1,Dance Monkey,Tones And I,0.825
29,Godzilla (feat. Juice WRLD),Eminem,0.808


There are many tracks with danceability above 0.7. These songs are likely easier to dance to because their danceability scores are high.

### Tracks With Danceability Below 0.4

In [31]:
danceability_below_04 = df[df["danceability"] < 0.4][["track_name", "artist", "danceability"]].sort_values(by="danceability", ascending=True)
danceability_below_04

,track_name,artist,danceability
44,lovely (with Khalid),Billie Eilish,0.351


Only one track has danceability below 0.4: `lovely (with Khalid)` by Billie Eilish. This shows that even a less danceable song can still become very popular.

### Tracks With Loudness Above -5

In [32]:
loudness_above_minus_5 = df[df["loudness"] > -5][["track_name", "artist", "loudness"]].sort_values(by="loudness", ascending=False)
loudness_above_minus_5

,track_name,artist,loudness
10,Tusa,KAROL G,-3.280
40,goosebumps,Travis Scott,-3.370
31,Break My Heart,Dua Lipa,-3.434
38,Hawái,Maluma,-3.454
12,Circles,Post Malone,-3.497
23,Mood (feat. iann dior),24kGoldn,-3.558
21,Adore You,Harry Styles,-3.675
49,SICKO MODE,Travis Scott,-3.714
48,Physical,Dua Lipa,-3.756
35,Rain On Me (with Ariana Grande),Lady Gaga,-3.764


These tracks have loudness above -5. Because loudness values are negative, numbers closer to zero mean louder songs.

### Tracks With Loudness Below -8

In [33]:
loudness_below_minus_8 = df[df["loudness"] < -8][["track_name", "artist", "loudness"]].sort_values(by="loudness", ascending=True)
loudness_below_minus_8

,track_name,artist,loudness
24,everything i wanted,Billie Eilish,-14.454
26,bad guy,Billie Eilish,-10.965
44,lovely (with Khalid),Billie Eilish,-10.109
47,If the World Was Ending - feat. Julia Michaels,JP Saxe,-10.086
15,Toosie Slide,Drake,-8.820
7,death bed (coffee for your head),Powfu,-8.765
36,HIGHEST IN THE ROOM,Travis Scott,-8.764
8,Falling,Trevor Daniel,-8.756
20,Savage Love (Laxed - Siren Beat),Jawsh 685,-8.520


These tracks are quieter compared with the rest of the dataset. Billie Eilish appears several times in this list, which suggests that quieter production can still perform well in the Top 50.

### Genres Represented in the Top 50

In [34]:
total_genres = df["genre"].nunique()
genre_counts = df["genre"].value_counts()
genres_with_one_song = genre_counts[genre_counts == 1]

print("Total genres:", total_genres)
genres_with_one_song

Total genres: 16


genre
Nu-disco                              1
R&B/Hip-Hop alternative               1
Pop/Soft Rock                         1
Pop rap                               1
Hip-Hop/Trap                          1
Dance-pop/Disco                       1
Disco-pop                             1
Dreampop/Hip-Hop/R&B                  1
Alternative/reggaeton/experimental    1
Chamber pop                           1
Name: count, dtype: int64

There are 16 genres represented in the Top 50. Several genres have only one song, which means they appear in the dataset but are not common in this Top 50 list.

### Comparison of Selected Genres

The assignment asks for a comparison between Pop, Hip-Hop/Rap, Dance/Electronic, and Alternative/Indie for danceability, loudness, and acousticness.

In [35]:
selected_genres = ["Pop", "Hip-Hop/Rap", "Dance/Electronic", "Alternative/Indie"]

selected_genre_comparison = (
    df[df["genre"].isin(selected_genres)]
    .groupby("genre")[["danceability", "loudness", "acousticness"]]
    .mean()
    .loc[selected_genres]
)

selected_genre_comparison

,danceability,loudness,acousticness
genre,,,
Pop,0.677571,-6.460357,0.323843
Hip-Hop/Rap,0.765538,-6.917846,0.188741
Dance/Electronic,0.755000,-5.338000,0.099440
Alternative/Indie,0.661750,-5.421000,0.583500


Among these four genres, Hip-Hop/Rap has the highest average danceability. Dance/Electronic has the loudest average loudness score because its value is closest to zero. Alternative/Indie has the highest average acousticness, meaning it sounds more acoustic on average than the other selected genres.

### Correlation Categories

Here I separate correlations into strong positive relationships, strong negative relationships, and examples of features that are not strongly correlated.

In [36]:
numeric_columns = [
    "energy", "danceability", "loudness", "acousticness", "speechiness",
    "instrumentalness", "liveness", "valence", "tempo", "duration_minutes"
]

correlation_matrix = df[numeric_columns].corr()
correlation_pairs = []

for i, first_column in enumerate(numeric_columns):
    for second_column in numeric_columns[i + 1:]:
        correlation_pairs.append({
            "feature_1": first_column,
            "feature_2": second_column,
            "correlation": correlation_matrix.loc[first_column, second_column]
        })

correlation_pairs_df = pd.DataFrame(correlation_pairs)

strong_positive_correlations = correlation_pairs_df[correlation_pairs_df["correlation"] >= 0.7].sort_values(by="correlation", ascending=False)
strong_negative_correlations = correlation_pairs_df[correlation_pairs_df["correlation"] <= -0.6].sort_values(by="correlation", ascending=True)
not_strongly_correlated = correlation_pairs_df[correlation_pairs_df["correlation"].abs() < 0.1].sort_values(by="correlation")

print("Strong positive correlations:")
display(strong_positive_correlations)

print("Strong negative correlations:")
display(strong_negative_correlations)

print("Examples of features that are not strongly correlated:")
display(not_strongly_correlated.head(10))

Strong positive correlations:


,feature_1,feature_2,correlation
1,energy,loudness,0.79164


Strong negative correlations:


,feature_1,feature_2,correlation
2,energy,acousticness,-0.682479


Examples of features that are not strongly correlated:


,feature_1,feature_2,correlation
41,liveness,duration_minutes,-0.090188
35,instrumentalness,liveness,-0.087034
20,loudness,liveness,-0.069939
43,valence,duration_minutes,-0.039794
16,danceability,duration_minutes,-0.033763
39,liveness,valence,-0.033366
18,loudness,speechiness,-0.021693
12,danceability,instrumentalness,-0.017706
29,acousticness,duration_minutes,-0.010988
13,danceability,liveness,-0.006648


The strongest positive correlation is between energy and loudness. This means higher-energy songs also tend to be louder.

The strongest negative correlation is between energy and acousticness. This means more acoustic songs tend to have lower energy in this dataset.

Some feature pairs have correlations close to zero, such as danceability and liveness. This means those features do not have a strong relationship in this dataset.

Correlation does not prove cause and effect. It only shows whether two numeric values tend to move together.

## Suggestions for Improvement

This analysis could be improved in several ways:

- Use a larger dataset with more songs and more years, so the results are not based only on 50 tracks from 2020.
- Compare 2020 with other years to see whether music trends changed over time.
- Add charts to make artist, genre, and correlation patterns easier to understand visually.
- Group similar genres together, because the dataset has many small genre labels that appear only once.
- Include Spotify popularity scores or streaming counts if available, so the analysis can compare audio features with actual popularity levels.

These improvements would make the analysis stronger and more useful for business decisions.

## Review Questions

### What is a DataFrame in Pandas?

A DataFrame is a table in Pandas. It has rows and columns, similar to a spreadsheet. In this project, each row is one Spotify track, and each column describes something about the track, such as artist, genre, danceability, or loudness.

### What makes computation in Pandas fast?

Pandas is fast because it is built on top of efficient lower-level tools like NumPy. Instead of processing values one by one in regular Python loops, Pandas can perform many calculations on whole columns at once. This is called vectorized computation.

### What advantages do Pandas Series objects have over Python lists?

A Pandas Series is like one column from a DataFrame. Compared with a regular Python list, a Series has helpful data analysis tools built in, such as `.mean()`, `.value_counts()`, `.isnull()`, and filtering. This makes it easier to clean, summarize, and analyze data.